artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerConsumerImpl.java

```diff
@@ -526,7 +526,11 @@ public void proceedDeliver(MessageReference reference) throws Exception {
          }
       } finally {
          pendingDelivery.countDown();
-         callback.afterDelivery();
+         synchronized (lock) {
```

Comment:
Given that this can run even when closing it in play as it just blocks the close cleanup bit from clearing the callback until it has been called would it make more sense to just capture the callback in a local variable under lock and then execute it outside the synchronized block to avoid staling a task on the message queue's executor?

Reply:
That's a good idea. I pushed an update. Let me know what you think.

Reply:
The main downside would be the extensive use of the session callback variable throughout the ServerComsumerImpl code do now all become volatile reads so will incur some small performance penalty compared to the past, how much of a penalty would require testing to really gauge.